File: OrderServiceApi/ServiceRegistry/ServiceRegistryClient.cs
namespace OrderServiceApi.ServiceRegistry;

using System;
using System.Threading.Tasks;
using Dtos;
using RestSharp;

public class ServiceRegistryClient
{
    private readonly RestClient _client;

    public ServiceRegistryClient(IConfiguration configuration)
    {
        var baseUrl = configuration.GetValue<Uri>("SideCarUrl");
        _client = new RestClient(baseUrl);
    }

    public async Task RegisterServiceAsync(RegistrationRequest request)
    {
        var restRequest = new RestRequest("services", Method.Post)
            .AddJsonBody(request);

        var response = await _client.ExecuteAsync(restRequest);

        if (!response.IsSuccessful)
        {
            throw new Exception("Failed to register service: " + response.Content);
        }
    }

    public async Task<RegistrationResponse> GetServiceAsync(string serviceName)
    {
        var restRequest = new RestRequest("services")
            .AddQueryParameter("serviceName", serviceName);

        var response = await _client.ExecuteAsync<RegistrationResponse>(restRequest);

        if (!response.IsSuccessful)
        {
            throw new Exception("Failed to get service: " + response.Content);
        }

        return response.Data;
    }
}

File: InventoryServiceApi/serviceregistryclient.py
import requests
from typing import Any

class ServiceRegistryClient:
    def __init__(self, sidecar_url: str):
        if not sidecar_url.endswith('/'):
            sidecar_url += '/'
        self.base_url = sidecar_url

    def register_service(self, request_data: Any) -> None:
        url = f"{self.base_url}services"
        headers = {'Content-Type': 'application/json'}
        response = requests.post(url, json=request_data, headers=headers)

        if not response.ok:
            raise Exception(f"Failed to register service: {response.text}")
